### DISCUSSION TITLE
when using persist middleware, arrays complain about being never[];

### DISCUSSION THREAD

**Discussion by Twiggeh** [pmndrs/zustand#1812] [Q&A]
[CodeSandbox](https://codesandbox.io/p/sandbox/practical-lake-mso9c9?file=%2Fsrc%2FApp.tsx%3A1%2C1-25%2C3)

This appears to break TS: 

```ts
import "./App.css";
import { create } from "zustand";
import { persist } from "zustand/middleware";

type ConnectionStore = { connections: string[] | undefined };

const connectionStore = create<ConnectionStore>()(
  // complains, when there is any array
  // persist(
  // breaks the type for partialize
  persist<ConnectionStore>(
    () => {
      return { connections: [] };
    },
    {
      name: "connections",
      partialize: (c) => {
        if (c.connections)
          return {
            connections: c.connections,
          };
      },
    }
  )
);
```

I tried re-adding the type to `persist`, but it appears to miss a `Partialize` ? 

I checked other discussions and issues but I couldnt find anything

**Comment by dbritto-dev** [comment]
@Twiggeh the state itself shouldn't be undefined you need to return at least an empty object

**Solution** 
Return an empty object if there's no connections or connections is falsy value

```ts
partialize: (c) => {
  if (c.connections) {
    return {
      connections: c.connections,
    };
  }
  return {}
}
```

**Reply by Twiggeh** [reply]
![20230517_20h28m11s_grim](https://github.com/pmndrs/zustand/assets/23149166/6abcc284-c81b-42df-b6bd-65a0344ff062)

The `never not assignable to string` is still there though

**Reply by dbritto-dev** [reply]
@Twiggeh   could you try with this?

```ts
const connectionStore = create<ConnectionStore>()(
  persist(() => ({ connections: [] as string[] }), {
    name: "connections",
    partialize: (s) => {
      if (s.connections) return { connections: s.connections };
      return {};
    },
  })
);
```

**Reply by Twiggeh** [reply]
@dbritto-dev the code snippet you wrote will still error, if you are going to typecast, you need to typecast `as string[] | undefined`. 

The requirement of the typecasting itself is an issue :/ It works perfectly fine without the persist middleware :(

**Reply by dbritto-dev** [reply]
> @dbritto-dev the code snippet you wrote will still error, if you are going to typecast, you need to typecast `as string[] | undefined`. 
> 
> The requirement of the typecasting itself is an issue :/ It works perfectly fine without the persist middleware :(

@Twiggeh you don't need to typecast to `as string[] | undefined` because you are typecasting the initial state so `as string[]` is completely OK.

About requiring to typecast when using is because we are passing in our initial state an empty array (there's no types that we can infer)

BTW, on persist each field is serialized/deserialized  (JSON.stringify/JSON.parse) so the types are expecting to have a value not an undefined

**Reply by JacobWeisenburger** [reply]
> @JacobWeisenburger could you try with this?
> 
> ```ts
> const connectionStore = create<ConnectionStore>()(
>   persist(() => ({ connections: [] as string[] }), {
>     name: "connections",
>     partialize: (s) => {
>       if (s.connections) return { connections: s.connections };
>       return {};
>     },
>   })
> );
> ```

Huh? I am not a part of this thread...

that's weird haha nvm

**Reply by Twiggeh** [reply]
@dbritto-dev 
> @Twiggeh you don't need to typecast to `as string[] | undefined` because you are typecasting the initial state so `as string[]` is completely OK.

You need to, otherwise the error doesn't go away ?

> About requiring to typecast when using is because we are passing in our initial state an empty array (there's no types that we can infer)

?

> BTW, on persist each field is serialized/deserialized (JSON.stringify/JSON.parse) so the types are expecting to have a value not an undefined

ok. 

The initial problem is that typecasts are needed in the first place, I think that might have gotten lost.
